dbms/src/Interpreters/Join.cpp

```diff
@@ -1926,40 +2004,51 @@ void Join::joinBlock(Block & block) const
 
     std::shared_lock lock(rwlock);
 
-    checkTypesOfKeys(block, sample_block_with_keys);
+    // if block != Block{}, it is for getHeader()
+    if (!block)
+    {
+        block = probe_process_infos[stream_index]->block;
+    }
```

Comment:
Can we spilt two function?
- the origin function used by `getHeader` and  `ExpressionAction`.
- the new function used by `readImpl`
```
Block Join::joinBlock(size_t stream_index) const
{
    Block block = probe_process_infos[stream_index]->block;
    joinBlock(block, stream_index);
    return block;
}
```

Reply:
> Can we spilt two function?
> 
> * the origin function used by `getHeader` and  `ExpressionAction`.
> * the new function used by `readImpl`
> 
> ```
> Block Join::joinBlock(size_t stream_index) const
> {
>     Block block = probe_process_infos[stream_index]->block;
>     joinBlock(block, stream_index);
>     return block;
> }
> ```

done.